### ISSUE TITLE
Dynamic public directories aren't given the same special treatement by vite as `publicDir` is

### ISSUE THREAD

**Issue by melink14** [nguyenbatranvan/vite-multiple-assets#41]
I don't think there's much we can do here but I wanted to open this in case it was worth documenting somewhere.

Currently, vite installs a special no-transform middleware for files in `publicDir`:
https://github.com/vitejs/vite/blob/d88d0004a8e891ca6026d356695e0b319caa7fce/packages/vite/src/node/server/index.ts#L882

There may be other occurences as well but there's at least that one.

During my migration, I saw one issue where a public css file was incorrectly transformed due to not being seen as public in the vite logic.

In that case, I was able to solve it by disabling the transform a different way but documenting it might save someone some time in the future. :)

Thanks!

**Comment by nguyenbatranvan** [comment]
> I don't think there's much we can do here but I wanted to open this in case it was worth documenting somewhere.
> 
> Currently, vite installs a special no-transform middleware for files in `publicDir`: https://github.com/vitejs/vite/blob/d88d0004a8e891ca6026d356695e0b319caa7fce/packages/vite/src/node/server/index.ts#L882
> 
> There may be other occurences as well but there's at least that one.
> 
> During my migration, I saw one issue where a public css file was incorrectly transformed due to not being seen as public in the vite logic.
> 
> In that case, I was able to solve it by disabling the transform a different way but documenting it might save someone some time in the future. :)
> 
> Thanks!

I haven’t quite grasped the issue you’re facing. Could you provide me with more details about it?

**Comment by nguyenbatranvan** [comment]
> Sorry, let me try to make it more concrete.
> 
> If I had a styles.css in a directory defined by this plugin and tried to fetch it naively in JS using a fetch request, it should be returned to me directly as a public file.
> 
> However, because vite doens't consider this plugin's files as public, it will apply a transform and turn the CSS into JS for importing.
> 
> It might actually be pretty rare since in the CSS case it can be avoided by setting the accept header correctly. In my case, my test runner plugin was causing the problem by not forwarding headers ([remcovaes/web-test-runner-vite-plugin#11](https://github.com/remcovaes/web-test-runner-vite-plugin/issues/11))
> 
> I only mention it here since it's a problem that doesn't occur when using `publicDir`.

Can you provide the code using JavaScript to load a CSS file?

**Comment by melink14** [comment]
I made a small repro in stackblitz: 

https://stackblitz.com/edit/vitejs-vite-vsmyr1me?file=src%2Fmain.ts

**Comment by nguyenbatranvan** [comment]
> I made a small repro in stackblitz:
> 
> https://stackblitz.com/edit/vitejs-vite-vsmyr1me?file=src%2Fmain.ts

I’ve done a quick test with the CSS. You can check the link below.

[https://stackblitz.com/edit/vitejs-vite-pxbktekg?file=vite.config.ts,src%2Fmain.ts,index.html,shared-2%2Fshared.css,.gitignore](https://stackblitz.com/edit/vitejs-vite-pxbktekg?file=vite.config.ts,src%2Fmain.ts,index.html,shared-2%2Fshared.css,.gitignore)

**Comment by nguyenbatranvan** [comment]
From the very beginning, my library was inspired by `micro-match` and `fast-glob`. If you want to exclude or include additional fields, please refer to the `dst` options of the library [here](https://nguyenbatranvan.github.io/vite-multiple-assets-doc/guide/index.html#optsdst).

**Comment by nguyenbatranvan** [comment]
Don't worry, I have a trick that might help you: modify the output path, so you won't need to change any files or folders in the `src` directory:
[https://stackblitz.com/edit/vitejs-vite-pmwzhigq?file=vite.config.ts,src%2Fmain.ts,dist%2Findex.html](https://stackblitz.com/edit/vitejs-vite-pmwzhigq?file=vite.config.ts,src%2Fmain.ts,dist%2Findex.html)

Let me know if there are any other challenges with your library that might be affecting my current work.

**Comment by melink14** [comment]
That's interesting, so as long as the output is not in the src directory it seems to work.

In my case the output file structure is the same as the input file structure so it seems that it will be trickier to fool vite.

Full source folder (extension is my `rootDir` in vite.config.ts:
```
extension
├── background.ts
├── clipboard.ts
├── configuration.ts
├── css
│   └── popup.css
├── data
│   ├── deinflect.dat
│   ├── dict.dat
│   ├── dict.idx
│   ├── kanji.dat
│   ├── names.dat
│   ├── names.idx
│   └── radicals.dat
├── data.ts
├── docs-annotate-canvas.ts
├── images
│   ├── ba.png
│   ├── icon128.png
│   └── icon48.png
├── manifest.json
├── offscreen.html
├── offscreen-setup.ts
├── offscreen.ts
├── options.html
├── options.ts
├── parse-word-dict-entry.ts
├── rikaichan.ts
├── rikaicontent.ts
├── test
│   ├── background_test.ts
│   ├── chrome_stubs.ts
│   ├── data_test.ts
│   ├── docs-annotate-canvas_test.ts
│   ├── e2e_visual_test.ts
│   ├── parse-word-dict-entry_test.ts
│   └── rikaicontent_test.ts
└── texttospeech.ts
```

My outdir is `../dist` (since it's relative to the root).

My `DynamicPublicDirectory` config is:
```ts
  DynamicPublicDirectory([
      '{\x01,data}/**',
      '{\x01,images}/**',
      '{\x01,css}/**',
      'manifest.json',
      '*.html',
    ]),
```

In `vite build` this works perfectly and copies all of the static files into my dist directory keeping the same file structure.

The problem in dev mode (during tests mainly) is that the fetches for CSS get transformed by default.

Thanks for taking a look!

**Comment by nguyenbatranvan** [comment]
> That's interesting, so as long as the output is not in the src directory it seems to work.
> 
> In my case the output file structure is the same as the input file structure so it seems that it will be trickier to fool vite.
> 
> Full source folder (extension is my `rootDir` in vite.config.ts:
> 
> ```
> extension
> ├── background.ts
> ├── clipboard.ts
> ├── configuration.ts
> ├── css
> │   └── popup.css
> ├── data
> │   ├── deinflect.dat
> │   ├── dict.dat
> │   ├── dict.idx
> │   ├── kanji.dat
> │   ├── names.dat
> │   ├── names.idx
> │   └── radicals.dat
> ├── data.ts
> ├── docs-annotate-canvas.ts
> ├── images
> │   ├── ba.png
> │   ├── icon128.png
> │   └── icon48.png
> ├── manifest.json
> ├── offscreen.html
> ├── offscreen-setup.ts
> ├── offscreen.ts
> ├── options.html
> ├── options.ts
> ├── parse-word-dict-entry.ts
> ├── rikaichan.ts
> ├── rikaicontent.ts
> ├── test
> │   ├── background_test.ts
> │   ├── chrome_stubs.ts
> │   ├── data_test.ts
> │   ├── docs-annotate-canvas_test.ts
> │   ├── e2e_visual_test.ts
> │   ├── parse-word-dict-entry_test.ts
> │   └── rikaicontent_test.ts
> └── texttospeech.ts
> ```
> 
> My outdir is `../dist` (since it's relative to the root).
> 
> My `DynamicPublicDirectory` config is:
> 
> ```ts
>   DynamicPublicDirectory([
>       '{\x01,data}/**',
>       '{\x01,images}/**',
>       '{\x01,css}/**',
>       'manifest.json',
>       '*.html',
>     ]),
> ```
> 
> In `vite build` this works perfectly and copies all of the static files into my dist directory keeping the same file structure.
> 
> The problem in dev mode (during tests mainly) is that the fetches for CSS get transformed by default.
> 
> Thanks for taking a look!

Alright, I might need some time to revisit this issue—it could take a little while! For now, are you able to run everything smoothly with my plugin?

**Comment by melink14** [comment]
Yeah, it's working great, and I just merged the migration to vite using your plugin. Thanks again!

I think this problem is pretty niche so it's probably not worth prioritizing; hopefully this thread is useful to any random person who runs into the same trouble in the future.

**Comment by nguyenbatranvan** [comment]
Once again, thank you so much for the issues you've contributed over the past time!

**Comment by nguyenbatranvan** [comment]
fixed `2.2.5`

**Comment by felixranesberger** [comment]
I don't know if this is related, but I have an issue where Vite tries to parse a compressed JS file that I load from a shared folder. I already tried to import the file using ?raw and it seemed to work, but suddenly the error started to show up again. 

```
[vite] Pre-transform error: ENOENT: no such file or directory, open '/styleguide-assets/client.js' (x65)
```

I load the asset from my html (which is also in the shared public folder) like this: 

```
<script type="module" src="/styleguide-assets/client.js?raw"></script>
```
